File: 08-09/Arquivo.java
import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.text.DecimalFormat;
import java.util.Scanner;

public class Arquivo
{
    // Objeto que formatará nossos números com as casas decimais corretas.
    public static final DecimalFormat FMT = new DecimalFormat("0.######");
    public static final String FN = "arquivoTesteJava.txt"; // Nome do arquivo.

    public static void main(String[] args)
    {
        // Abre o arquivo no modo de escrita.
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(FN))) {
            Scanner sc = new Scanner(System.in);
            final int numLines = sc.nextInt(); // Lê o número de linhas.

            for (int i = 0; i < numLines; ++i) {
                double input = sc.nextDouble(); // Lê os números na entrada.
                writer.write(formataDouble(input) + "\n"); // Salva no arquivo.
            }

            sc.close(); // Fecha o Scanner.
        } catch (IOException e) {
            e.printStackTrace();
        }

        // Abre o arquivo no modo de leitura.
        try (RandomAccessFile arq = new RandomAccessFile(FN, "r")) {
            // Itera sobre o arquivo de trás pra frente.
            for (long i = arq.length() - 1; i >= 0; --i) {
                arq.seek(i); // Move para a i-ésima posição.

                // Verifica se está num fim de linha ou no começo do arquivo.
                if (arq.readByte() == '\n' || i == 0) {
                    if (i != 0)
                        // Avança uma posição após '\n'.
                        arq.seek(i + 1);
                    else
                        // No começo do arquivo, não avança (reseta posição).
                        arq.seek(i);

                    // Lê um valor dessa linha.
                    String lin = arq.readLine();
                    // Confere se a linha existe e não é vazia.
                    if (lin != null && lin.length() != 0) {
                        double input = Double.parseDouble(lin); // Converte.
                        System.out.println(formataDouble(input)); // Formata.
                    }
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    // Converte um double para String, usando a formatação esperada no Verde.
    public static String formataDouble(double d)
    {
        return FMT.format(d);
    }
}

File: 08-09. Arquivo/arquivo.c
#include <stdio.h>

#define FN "arquivo_teste_c.txt" // Nome do arquivo.

int main(void)
{
	FILE *fd; // Objeto representando o arquivo.
	int num; // Número de linhas que serão processadas.

	scanf("%d", &num); // Lê o número de linhas.

	fd = fopen(FN, "w"); // Abre o arquivo em modo de escrita.
	for (int i = 0; i < num; ++i) {
		double input;
		scanf("%lf", &input); // Lê entrada.
		fprintf(fd, "\n%g", input); // Salva entrada no arquivo.
	}
	fclose(fd); // Fecha o arquivo.

	fd = fopen(FN, "r"); // Abre o arquivo em modo de leitura.

	// Move o cursor para o final do arquivo
	fseek(fd, 0, SEEK_END);

	// Itera sobre o arquivo de trás pra frente.
	for (long long i = ftell(fd) - 1; i >= 0; --i) {
		fseek(fd, i, SEEK_SET); // Move para a i-ésima posição.

		// Verifica se está num fim de linha ou no começo do arquivo.
		if (fgetc(fd) == '\n' || i == 0) {
			double input;
			if (i != 0)
				// Avança uma posição após '\n'.
				fseek(fd, i + 1, SEEK_SET);
			else
				// No começo do arquivo, não avança.
				fseek(fd, i, SEEK_SET);

			fscanf(fd, "%lf", &input); // Lê um valor dessa linha.
			printf("%g\n", input); // Mostra o valor.
		}
	}

	fclose(fd); // Fecha o arquivo.
	return 0;
}
